packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,230 @@
+
+import { Script } from "forge-std/Script.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+
+import { L2StandardBridge } from "src/L2/L2StandardBridge.sol";
+import { L2CrossDomainMessenger } from "src/L2/L2CrossDomainMessenger.sol";
+import { SequencerFeeVault } from "src/L2/SequencerFeeVault.sol";
+import { FeeVault } from "src/universal/FeeVault.sol";
+import { OptimismMintableERC20Factory } from "src/universal/OptimismMintableERC20Factory.sol";
+import { GovernanceToken } from "src/governance/GovernanceToken.sol";
+
+interface IInitializable {
+    function initialize() external;
+}
+
+// @title
+contract L2Genesis is Script {
+    uint256 constant PROXY_COUNT = 2048;
+    uint256 constant PRECOMPILE_COUNT = 256;
+
+    /// @notice The storage slot that holds the address of a proxy implementation.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.implementation')) - 1)`
+    bytes32 internal constant PROXY_IMPLEMENTATION_ADDRESS =
+        0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc;
+
+    /// @notice The storage slot that holds the address of the owner.
+    /// @dev `bytes32(uint256(keccak256('eip1967.proxy.admin')) - 1)`
+    bytes32 internal constant PROXY_ADMIN_ADDRESS = 0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103;
+
+    string outfile;
+
+    function setUp() external {
+        outfile = string.concat(vm.projectRoot(), "/genesis.json");
+    }
+
+    function run() public {
+        _setPrecompiles();
+        _setProxies();
+        _setImplementations();
+
+        vm.dumpState(outfile);
+    }
+
+    /// @notice Give all of the precompiles 1 wei so that they are
+    ///         not considered empty accounts.
+    function _setPrecompiles() internal {
+        for (uint256 i; i < PRECOMPILE_COUNT; i++) {
+            vm.deal(address(uint160(i)), 1);
+        }
+    }
+
+    /// @dev Set up the accounts that correspond to the predeploys.
+    ///      The Proxy bytecode should be set. All proxied predeploys should have
+    ///      the 1967 admin slot set to the ProxyAdmin predeploy. All defined predeploys
+    ///      should have their implementations set.
+    function _setProxies() internal {
+        bytes memory code = vm.getDeployedCode("Proxy.sol:Proxy");
+        uint160 prefix = uint160(0x420) << 148;
+        for (uint256 i = 0; i < PROXY_COUNT; i++) {
+            address addr = address(prefix | uint160(i));
+            if (_notProxied(addr)) {
+                continue;
+            }
+
+            vm.etch(addr, code);
+            vm.store(addr, PROXY_ADMIN_ADDRESS, bytes32(uint256(uint160(Predeploys.PROXY_ADMIN))));
+
+            if (_isDefinedPredeploy(addr)) {
+                address implementation = _predeployToCodeNamespace(addr);
+                vm.store(addr, PROXY_IMPLEMENTATION_ADDRESS, bytes32(uint256(uint160(implementation))));
+            }
+        }
+    }
+
+    /// @dev
+    /// 2 options
+    ///  - use getDeployedCode
```

Comment:
`getDeployedCode` is safe when there are no immutables in the contract. Not sure if there is a way to reconcile that with foundry, ie encoding them as bytes as a second arg to `getDeployedCode` could work but is hacky, not sure if forge supports that. So if it has immutables, must use `new` to set the code. We want to avoid using `new` as much as possible because it 1) slows down compile time and 2) creates accounts that we need to empty out (nonce, possible storage, bytecode) and then move that stuff to the proper place in the specified address (`0xc0d3...` implementation address)